src/pages/workspaces/workspace/notebooks/AppLauncher.js

```diff
@@ -0,0 +1,126 @@
+import _ from 'lodash/fp'
+import { Fragment, useRef, useState } from 'react'
+import { div, h, iframe } from 'react-hyperscript-helpers'
+import * as breadcrumbs from 'src/components/breadcrumbs'
+import { PlaygroundHeader, StatusMessage } from 'src/components/cluster-common'
+import { Link, spinnerOverlay } from 'src/components/common'
+import { NewClusterModal } from 'src/components/NewClusterModal'
+import { Ajax } from 'src/libs/ajax'
+import * as Auth from 'src/libs/auth'
+import { handleNonRunningCluster } from 'src/libs/cluster-utils'
+import { reportError, withErrorReporting } from 'src/libs/error'
+import * as Utils from 'src/libs/utils'
+import { wrapWorkspace } from 'src/pages/workspaces/workspace/WorkspaceContainer'
+
+
+const AppLauncher = _.flow(
+  Utils.forwardRefWithName('AppLauncher'),
+  wrapWorkspace({
+    breadcrumbs: props => breadcrumbs.commonPaths.workspaceDashboard(props),
+    title: _.get('app'),
+    activeTab: 'notebooks'
+  })
+)(({ namespace, refreshClusters, cluster, app }, ref) => {
+  const signal = Utils.useCancellation()
+  const { Clusters } = Ajax(signal)
+  const [cookieReady, setCookieReady] = useState(false)
+  const [shownCreate, setShownCreate] = useState(false)
+  const [busy, setBusy] = useState(false)
+  const getCluster = Utils.useGetter(cluster)
+
+  const cookieTimeout = useRef()
+  const iframeRef = useRef()
+
+  Utils.useOnMount(() => {
+    const refreshCookie = async () => {
+      const { expires_in: expiresInSec } = await Auth.reloadAuthToken() // give us max time until expiry, should be 1 hour
+
+      cookieTimeout.current = setTimeout(refreshCookie, (expiresInSec - (15 * 60)) * 1000) // refresh 15 mins early
+
+      await Clusters.notebooks(namespace, getCluster().clusterName).setCookie()
+    }
+
+    const startCluster = async () => {
+      while (true) {
+        try {
+          await refreshClusters()
+          const cluster = getCluster()
+          const status = cluster?.status
+
+          if (status === 'Running') {
+            await refreshCookie()
+            setCookieReady(true)
+            iframeRef.current.onload = function() { this.contentWindow.focus() }
+            return
```

Comment:
Since this currently only runs on mount at the page level, this flow will not be repeated if the cluster gets stopped and then restarted without leaving the page. We might want to do something like the `NotebookLauncher` page, where the iframe is pushed into a sub-component along with its setup code. That way, we can leverage the mount/unmount cycle to perform these steps each time.